File: packages/sdk-node/tests/ip-filter.test.ts
import { describe, it, expect } from 'vitest';
import {
  ipToNumber,
  parseCidr,
  ipMatchesCidr,
  ipMatchesAny,
  shouldAllowIp,
  checkIpFilter,
  type IpFilterConfig,
} from '../src/utils/ip-filter';

describe('IP Filter Utilities', () => {
  describe('ipToNumber', () => {
    it('should parse valid IPv4 addresses', () => {
      expect(ipToNumber('0.0.0.0')).toBe(0);
      expect(ipToNumber('255.255.255.255')).toBe(4294967295);
      expect(ipToNumber('192.168.1.1')).toBe(3232235777);
      expect(ipToNumber('10.0.0.1')).toBe(167772161);
    });

    it('should return null for invalid IPs', () => {
      expect(ipToNumber('invalid')).toBe(null);
      expect(ipToNumber('256.1.1.1')).toBe(null);
      expect(ipToNumber('1.2.3')).toBe(null);
      expect(ipToNumber('1.2.3.4.5')).toBe(null);
      expect(ipToNumber('')).toBe(null);
      expect(ipToNumber('a.b.c.d')).toBe(null);
      expect(ipToNumber('1.2.3.-1')).toBe(null);
    });
  });

  describe('parseCidr', () => {
    it('should parse CIDR notation', () => {
      const result = parseCidr('192.168.1.0/24');
      expect(result).not.toBe(null);
      expect(result?.mask).toBe(4294967040); // 255.255.255.0
    });

    it('should parse single IP as /32', () => {
      const result = parseCidr('192.168.1.1');
      expect(result).not.toBe(null);
      expect(result?.mask).toBe(4294967295); // 255.255.255.255
    });

    it('should handle /0 (all IPs)', () => {
      const result = parseCidr('0.0.0.0/0');
      expect(result).not.toBe(null);
      expect(result?.mask).toBe(0);
    });

    it('should return null for invalid CIDR', () => {
      expect(parseCidr('invalid/24')).toBe(null);
      expect(parseCidr('192.168.1.0/33')).toBe(null);
      expect(parseCidr('192.168.1.0/-1')).toBe(null);
    });
  });

  describe('ipMatchesCidr', () => {
    it('should match exact IP', () => {
      expect(ipMatchesCidr('192.168.1.1', '192.168.1.1')).toBe(true);
      expect(ipMatchesCidr('192.168.1.2', '192.168.1.1')).toBe(false);
    });

    it('should match /24 subnet', () => {
      expect(ipMatchesCidr('192.168.1.0', '192.168.1.0/24')).toBe(true);
      expect(ipMatchesCidr('192.168.1.1', '192.168.1.0/24')).toBe(true);
      expect(ipMatchesCidr('192.168.1.255', '192.168.1.0/24')).toBe(true);
      expect(ipMatchesCidr('192.168.2.0', '192.168.1.0/24')).toBe(false);
    });

    it('should match /16 subnet', () => {
      expect(ipMatchesCidr('192.168.0.0', '192.168.0.0/16')).toBe(true);
      expect(ipMatchesCidr('192.168.255.255', '192.168.0.0/16')).toBe(true);
      expect(ipMatchesCidr('192.169.0.0', '192.168.0.0/16')).toBe(false);
    });

    it('should match /8 subnet', () => {
      expect(ipMatchesCidr('10.0.0.1', '10.0.0.0/8')).toBe(true);
      expect(ipMatchesCidr('10.255.255.255', '10.0.0.0/8')).toBe(true);
      expect(ipMatchesCidr('11.0.0.0', '10.0.0.0/8')).toBe(false);
    });

    it('should match /32 (single IP)', () => {
      expect(ipMatchesCidr('203.0.113.50', '203.0.113.50/32')).toBe(true);
      expect(ipMatchesCidr('203.0.113.51', '203.0.113.50/32')).toBe(false);
    });

    it('should match /0 (all IPs)', () => {
      expect(ipMatchesCidr('1.2.3.4', '0.0.0.0/0')).toBe(true);
      expect(ipMatchesCidr('255.255.255.255', '0.0.0.0/0')).toBe(true);
    });

    it('should return false for invalid IPs', () => {
      expect(ipMatchesCidr('invalid', '192.168.1.0/24')).toBe(false);
      expect(ipMatchesCidr('192.168.1.1', 'invalid/24')).toBe(false);
    });
  });

  describe('ipMatchesAny', () => {
    it('should match if IP is in any entry', () => {
      const list = ['192.168.1.0/24', '10.0.0.0/8', '203.0.113.50'];
      expect(ipMatchesAny('192.168.1.100', list)).toBe(true);
      expect(ipMatchesAny('10.50.25.1', list)).toBe(true);
      expect(ipMatchesAny('203.0.113.50', list)).toBe(true);
    });

    it('should return false if IP is not in any entry', () => {
      const list = ['192.168.1.0/24', '10.0.0.0/8'];
      expect(ipMatchesAny('172.16.0.1', list)).toBe(false);
      expect(ipMatchesAny('8.8.8.8', list)).toBe(false);
    });

    it('should return false for empty list', () => {
      expect(ipMatchesAny('192.168.1.1', [])).toBe(false);
    });
  });

  describe('shouldAllowIp', () => {
    describe('blocklist mode', () => {
      it('should block IPs in blocklist', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'blocklist',
          blocklist: ['192.168.1.0/24', '203.0.113.0/24'],
        };
        const result = shouldAllowIp('192.168.1.50', config);
        expect(result.allowed).toBe(false);
        expect(result.reason).toBe('blocklist');
      });

      it('should allow IPs not in blocklist', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'blocklist',
          blocklist: ['192.168.1.0/24'],
        };
        const result = shouldAllowIp('10.0.0.1', config);
        expect(result.allowed).toBe(true);
        expect(result.reason).toBe('default');
      });

      it('should allow all if blocklist is empty', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'blocklist',
          blocklist: [],
        };
        const result = shouldAllowIp('192.168.1.1', config);
        expect(result.allowed).toBe(true);
      });
    });

    describe('allowlist mode', () => {
      it('should allow IPs in allowlist', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'allowlist',
          allowlist: ['10.0.0.0/8', '192.168.0.0/16'],
        };
        const result = shouldAllowIp('10.0.0.50', config);
        expect(result.allowed).toBe(true);
        expect(result.reason).toBe('allowlist');
      });

      it('should block IPs not in allowlist', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'allowlist',
          allowlist: ['10.0.0.0/8'],
        };
        const result = shouldAllowIp('192.168.1.1', config);
        expect(result.allowed).toBe(false);
        expect(result.reason).toBe('not_in_allowlist');
      });

      it('should block all if allowlist is empty', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'allowlist',
          allowlist: [],
        };
        const result = shouldAllowIp('10.0.0.1', config);
        expect(result.allowed).toBe(false);
      });
    });

    describe('both mode', () => {
      it('should allow if in allowlist even if in blocklist range', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'both',
          allowlist: ['10.0.0.1'],
          blocklist: ['10.0.0.0/24'],
        };
        const result = shouldAllowIp('10.0.0.1', config);
        expect(result.allowed).toBe(true);
        expect(result.reason).toBe('allowlist');
      });

      it('should block if in blocklist and not in allowlist', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'both',
          allowlist: ['10.0.0.1'],
          blocklist: ['10.0.0.0/24'],
        };
        const result = shouldAllowIp('10.0.0.2', config);
        expect(result.allowed).toBe(false);
        expect(result.reason).toBe('blocklist');
      });

      it('should allow if not in allowlist or blocklist', () => {
        const config: IpFilterConfig = {
          enabled: true,
          mode: 'both',
          allowlist: ['10.0.0.1'],
          blocklist: ['192.168.1.0/24'],
        };
        const result = shouldAllowIp('8.8.8.8', config);
        expect(result.allowed).toBe(true);
        expect(result.reason).toBe('default');
      });
    });
  });

  describe('checkIpFilter', () => {
    it('should use custom filter when provided', async () => {
      const config: IpFilterConfig = {
        enabled: true,
        mode: 'blocklist',
        blocklist: ['192.168.1.0/24'],
        customFilter: (ip) => {
          // Custom rule: block all IPs starting with "10."
          if (ip.startsWith('10.')) return false;
          return undefined; // Defer to list-based filtering
        },
      };

      // Custom filter blocks 10.x.x.x
      const result1 = await checkIpFilter('10.0.0.1', config, { path: '/test' });
      expect(result1.allowed).toBe(false);
      expect(result1.reason).toBe('custom');

      // Custom filter defers, blocklist blocks 192.168.1.x
      const result2 = await checkIpFilter('192.168.1.50', config, { path: '/test' });
      expect(result2.allowed).toBe(false);
      expect(result2.reason).toBe('blocklist');

      // Custom filter defers, not in blocklist
      const result3 = await checkIpFilter('8.8.8.8', config, { path: '/test' });
      expect(result3.allowed).toBe(true);
      expect(result3.reason).toBe('default');
    });

    it('should handle async custom filter', async () => {
      const config: IpFilterConfig = {
        enabled: true,
        mode: 'blocklist',
        customFilter: async (ip) => {
          // Simulate async check (e.g., database lookup)
          await new Promise((resolve) => setTimeout(resolve, 1));
          return ip === '8.8.8.8' ? false : undefined;
        },
      };

      const result = await checkIpFilter('8.8.8.8', config, { path: '/test' });
      expect(result.allowed).toBe(false);
      expect(result.reason).toBe('custom');
    });

    it('should work without custom filter', async () => {
      const config: IpFilterConfig = {
        enabled: true,
        mode: 'blocklist',
        blocklist: ['192.168.1.0/24'],
      };

      const result = await checkIpFilter('192.168.1.50', config, { path: '/test' });
      expect(result.allowed).toBe(false);
      expect(result.reason).toBe('blocklist');
    });
  });
});

File: packages/sdk-go/ip_filter_test.go
package pocketping

import (
	"context"
	"testing"
)

func TestIpMatchesCidr(t *testing.T) {
	tests := []struct {
		name     string
		ip       string
		cidr     string
		expected bool
	}{
		// Exact IP match
		{"exact match", "192.168.1.1", "192.168.1.1", true},
		{"exact no match", "192.168.1.2", "192.168.1.1", false},

		// /24 subnet
		{"/24 match start", "192.168.1.0", "192.168.1.0/24", true},
		{"/24 match middle", "192.168.1.1", "192.168.1.0/24", true},
		{"/24 match end", "192.168.1.255", "192.168.1.0/24", true},
		{"/24 no match", "192.168.2.0", "192.168.1.0/24", false},

		// /16 subnet
		{"/16 match", "192.168.0.0", "192.168.0.0/16", true},
		{"/16 match high", "192.168.255.255", "192.168.0.0/16", true},
		{"/16 no match", "192.169.0.0", "192.168.0.0/16", false},

		// /8 subnet
		{"/8 match low", "10.0.0.1", "10.0.0.0/8", true},
		{"/8 match high", "10.255.255.255", "10.0.0.0/8", true},
		{"/8 no match", "11.0.0.0", "10.0.0.0/8", false},

		// /32 (single IP)
		{"/32 match", "203.0.113.50", "203.0.113.50/32", true},
		{"/32 no match", "203.0.113.51", "203.0.113.50/32", false},

		// /0 (all IPs)
		{"/0 match any", "1.2.3.4", "0.0.0.0/0", true},
		{"/0 match all", "255.255.255.255", "0.0.0.0/0", true},

		// Invalid inputs
		{"invalid ip", "invalid", "192.168.1.0/24", false},
		{"invalid cidr", "192.168.1.1", "invalid/24", false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ipMatchesCidr(tt.ip, tt.cidr)
			if result != tt.expected {
				t.Errorf("ipMatchesCidr(%q, %q) = %v, want %v", tt.ip, tt.cidr, result, tt.expected)
			}
		})
	}
}

func TestIpMatchesAny(t *testing.T) {
	entries := []string{"192.168.1.0/24", "10.0.0.0/8", "203.0.113.50"}

	tests := []struct {
		name     string
		ip       string
		expected bool
	}{
		{"matches first range", "192.168.1.100", true},
		{"matches second range", "10.50.25.1", true},
		{"matches exact ip", "203.0.113.50", true},
		{"no match", "172.16.0.1", false},
		{"no match 2", "8.8.8.8", false},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			result := ipMatchesAny(tt.ip, entries)
			if result != tt.expected {
				t.Errorf("ipMatchesAny(%q, entries) = %v, want %v", tt.ip, result, tt.expected)
			}
		})
	}

	// Empty list test
	t.Run("empty list", func(t *testing.T) {
		result := ipMatchesAny("192.168.1.1", []string{})
		if result != false {
			t.Errorf("ipMatchesAny with empty list = %v, want false", result)
		}
	})
}

func TestShouldAllowIP(t *testing.T) {
	t.Run("blocklist mode blocks", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeBlocklist,
			Blocklist: []string{"192.168.1.0/24", "203.0.113.0/24"},
		}
		result := ShouldAllowIP("192.168.1.50", config)
		if result.Allowed != false || result.Reason != IpFilterReasonBlocklist {
			t.Errorf("Expected blocked with blocklist reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("blocklist mode allows", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeBlocklist,
			Blocklist: []string{"192.168.1.0/24"},
		}
		result := ShouldAllowIP("10.0.0.1", config)
		if result.Allowed != true || result.Reason != IpFilterReasonDefault {
			t.Errorf("Expected allowed with default reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("allowlist mode allows", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeAllowlist,
			Allowlist: []string{"10.0.0.0/8", "192.168.0.0/16"},
		}
		result := ShouldAllowIP("10.0.0.50", config)
		if result.Allowed != true || result.Reason != IpFilterReasonAllowlist {
			t.Errorf("Expected allowed with allowlist reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("allowlist mode blocks", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeAllowlist,
			Allowlist: []string{"10.0.0.0/8"},
		}
		result := ShouldAllowIP("192.168.1.1", config)
		if result.Allowed != false || result.Reason != IpFilterReasonNotInAllowlist {
			t.Errorf("Expected blocked with not_in_allowlist reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("both mode allowlist priority", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeBoth,
			Allowlist: []string{"10.0.0.1"},
			Blocklist: []string{"10.0.0.0/24"},
		}
		result := ShouldAllowIP("10.0.0.1", config)
		if result.Allowed != true || result.Reason != IpFilterReasonAllowlist {
			t.Errorf("Expected allowed with allowlist reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("both mode blocklist applies", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeBoth,
			Allowlist: []string{"10.0.0.1"},
			Blocklist: []string{"10.0.0.0/24"},
		}
		result := ShouldAllowIP("10.0.0.2", config)
		if result.Allowed != false || result.Reason != IpFilterReasonBlocklist {
			t.Errorf("Expected blocked with blocklist reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("nil config allows", func(t *testing.T) {
		result := ShouldAllowIP("192.168.1.1", nil)
		if result.Allowed != true {
			t.Errorf("Expected allowed for nil config, got allowed=%v", result.Allowed)
		}
	})
}

func TestCheckIPFilter(t *testing.T) {
	t.Run("custom filter blocks", func(t *testing.T) {
		falseVal := false
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeBlocklist,
			Blocklist: []string{"192.168.1.0/24"},
			CustomFilter: func(ip string, requestInfo map[string]interface{}) *bool {
				// Block all IPs starting with "10."
				if len(ip) >= 3 && ip[:3] == "10." {
					return &falseVal
				}
				return nil // Defer to list-based
			},
		}

		result := CheckIPFilter(context.Background(), "10.0.0.1", config, nil)
		if result.Allowed != false || result.Reason != IpFilterReasonCustom {
			t.Errorf("Expected blocked with custom reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("custom filter allows", func(t *testing.T) {
		trueVal := true
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeBlocklist,
			Blocklist: []string{"192.168.1.0/24"},
			CustomFilter: func(ip string, requestInfo map[string]interface{}) *bool {
				if ip == "8.8.8.8" {
					return &trueVal
				}
				return nil
			},
		}

		result := CheckIPFilter(context.Background(), "8.8.8.8", config, nil)
		if result.Allowed != true || result.Reason != IpFilterReasonCustom {
			t.Errorf("Expected allowed with custom reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("custom filter defers", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   true,
			Mode:      IpFilterModeBlocklist,
			Blocklist: []string{"192.168.1.0/24"},
			CustomFilter: func(ip string, requestInfo map[string]interface{}) *bool {
				return nil // Defer
			},
		}

		result := CheckIPFilter(context.Background(), "192.168.1.50", config, nil)
		if result.Allowed != false || result.Reason != IpFilterReasonBlocklist {
			t.Errorf("Expected blocked with blocklist reason, got allowed=%v, reason=%v", result.Allowed, result.Reason)
		}
	})

	t.Run("disabled config allows", func(t *testing.T) {
		config := &IpFilterConfig{
			Enabled:   false,
			Mode:      IpFilterModeBlocklist,
			Blocklist: []string{"192.168.1.0/24"},
		}

		result := CheckIPFilter(context.Background(), "192.168.1.50", config, nil)
		if result.Allowed != true {
			t.Errorf("Expected allowed for disabled config, got allowed=%v", result.Allowed)
		}
	})
}
